ibis/expr/operations/tvf.py

```diff
@@ -0,0 +1,121 @@
+from __future__ import annotations
+
+from abc import abstractmethod
+from enum import Enum
+from typing import Optional
+
+from public import public
+
+import ibis.expr.datatypes as dt
+from ibis.expr.operations.core import Column, Scalar  # noqa: TCH001
+from ibis.expr.operations.relations import Relation
+from ibis.expr.schema import Schema
+
+
+@public
+class TableValuedFunctionTypes(str, Enum):
+    TUMBLE = "TUMBLE"
+    HOP = "HOP"
+    CUMULATE = "CUMULATE"
+
+    @classmethod
+    def list(cls):
+        return [c.value for c in cls]
+
+
+@public
+class TableValuedFunction(Relation):
+    table: Relation
+
+    def __init__(self, table, **kwargs):
+        super().__init__(table=table, **kwargs)
+
+    @property
+    @abstractmethod
+    def function_type(self) -> str:
+        """Abstract class attribute.
+
+        This makes TableValuedFunction and WindowTVF abstract classes.
```

Comment:
Relation is already an abstract class requiring a `schema` attribute. Is the `function_type` attribute desired? Looks like the type uniquely determines the value of it.

Reply:
`schema` is implemented on `WindowTVF`, so `TableValuedFunction` is an abstract class but `WindowTVF` isn't. The intention of having `function_type` is mostly just to make `WindowTVF` an abstract class, but I don't feel strongly about this implementation. We can certainly remove it.